warehouse/query-core/src/main/java/datawave/query/index/lookup/TupleToRange.java

```diff
@@ -35,79 +35,120 @@
-                // Technically, we don't want to be inclusive of the start key,
-                // however if we mark the startKey as non-inclusive, when we create
-                // the fi\x00 range in IndexIterator, we lost the context of "do we
-                // want a single event" or "did we get restarted and this is the last
-                // event we returned.
-                Range r = new Range(start, true, end, false);
-                
-                if (log.isTraceEnabled())
-                    log.trace(queryNode + " " + uid.getNode());
-                
-                // don't really want log statement if uid.getNode is null
-                if (log.isTraceEnabled() && null != uid.getNode()) {
-                    
-                    // query node can be null in this case
-                    log.trace("Building " + r + " from " + (null == queryNode ? "NoQueryNode" : JexlStringBuildingVisitor.buildQuery(queryNode)) + " actually "
-                                    + JexlStringBuildingVisitor.buildQuery(uid.getNode()));
-                }
-                
-                ranges.add(new QueryPlan(uid.getNode(), r));
-            }
-            return ranges.iterator();
-        }
-        // else if this is a shard, then range from <shard> to <shard>\x00
-        else if (tuple.first().indexOf('_') >= 0) {
-            JexlNode myNode = queryNode;
-            if (ii.getNode() != null) {
-                myNode = ii.getNode();
-            }
-            if (log.isTraceEnabled() && null != myNode)
-                log.trace("Building shard " + new Range(tuple.first(), true, tuple.first() + NULL_BYTE_STRING, false) + " From "
-                                + (null == myNode ? "NoQueryNode" : JexlStringBuildingVisitor.buildQuery(myNode)));
+        if (isDocumentRange(indexInfo)) {
+            
+            return createDocumentRanges(queryNode, shard, indexInfo);
+            
+        } else if (isShardRange(shard)) {
+            
+            return createShardRange(queryNode, shard, indexInfo);
+            
+        } else {
             
-            return Collections.singleton(new QueryPlan(myNode, new Range(tuple.first(), true, tuple.first() + NULL_BYTE_STRING, false))).iterator();
+            return createDayRange(queryNode, shard, indexInfo);
         }
-        // else assume this a day range, then range from <day> to <day>\xff...
-        else {
+    }
+    
+    /**
+     * Building document ranges is only possible if the IndexInfo object contains document ids.
+     *
+     * @param indexInfo
+     *            - object built from matches in the index.
+     * @return - true if we can build document range(s).
+     */
+    public boolean isDocumentRange(IndexInfo indexInfo) {
+        return !indexInfo.uids().isEmpty();
+    }
+    
+    /**
+     *
+     * @param shard
+     * @return - true if the shard string is a shard range
+     */
+    public boolean isShardRange(String shard) {
```

Comment:
Arguably, we know exactly where the "_" should be, no? If we do, should we be checking that the character at that index, instead of running an indexOf?

Reply:
In theory we do know, however the standard seems to be an indexOf() call.